lib/rpmsg/rpmsg_virtio.c

```diff
@@ -23,6 +23,27 @@
 /* Time to wait - In multiple of 10 msecs. */
 #define RPMSG_TICKS_PER_INTERVAL                10
 
+static void *
+rpmsg_virtio_shm_pool_get_buffer(struct rpmsg_virtio_shm_pool *shpool,
+				 size_t size)
+{
+	void *buffer;
+
+	if(shpool->avail < size)
```

Comment:
what about creating a generic rpmsg_dev_ops structure with associted  APi register some app ops that will modify default behaviour?
rpmsg_virtio_shm_pool_get_buffer should be part of this ops structure an we could simply extend ops on need in future (e.g.  log).
Then in rpmsg_virtio_shm_pool_get_buffer we would have 
if (ops->get_buffer)
      ops->get_buffer (...);
else
   /* do the default action */

Reply:
my remark here is more to introduce ops functions ( similar to kernel frameworks that offer ops for drivers that want a specific management of a service). Instead of updating API each  time we need a specific treatment by application, we can define ops structures that list all this kind of pointers. Perhaps we need rpmsg ops , rpmsg_virtio ops, remote proc ops and libmetal ops strutures...

Reply:
i agree with more ops, but just whether it should be rpmsg_virtio ops, rpmsg ops or shared buffered ops itself. it is more closer to rpmsg_virtio, however, i more inclined to shared buffers itself, as the shared buffers ops needs its own private data. And thus, i wrap the shared buffer ops + its data into the shared buffer pool struct